extension/intents/nicknames/routineExecutor.js

```diff
@@ -0,0 +1,102 @@
+/* globals log */
+
+import * as intentRunner from "../../background/intentRunner.js";
+
+export class RoutineExecutor {
+  constructor(
+    routineName,
+    subcommands,
+    index = 0,
+    states = [{}],
+    forIndex = 0
+  ) {
+    this.routineName = routineName;
+    this.subcommands = subcommands;
+    this.index = index;
+    this.states = states;
+    this.forIndex = forIndex;
+  }
+
+  mapState(subcommand, state) {
+    const copySubcommand = JSON.parse(JSON.stringify(subcommand));
+    for (const slot in subcommand.slots) {
+      if (subcommand.slots[slot] in state) {
+        copySubcommand.slots[slot] = state[subcommand.slots[slot]];
+      }
+    }
+    return copySubcommand;
+  }
+
+  async runSubcommand(subcommand) {
+    let hadError = false;
+    let errorMessage = null;
+
+    log.info(
+      "  Running subintent",
+      subcommand,
+      subcommand.name,
+      subcommand.slots,
+      this.index
+    );
+
+    subcommand = this.mapState(subcommand, this.states[0]);
+    subcommand.routineExecutor = this;
+    subcommand.onError = message => {
+      hadError = true;
+      errorMessage = message;
+    };
+    await intentRunner.runIntent(subcommand);
+    if (hadError) {
+      log.info(
+        `Routine ${this.routineName} encountered error ${errorMessage} while running ${subcommand.utterance}`
+      );
+      return true;
+    }
+
+    return this._stop === true;
+  }
+
+  startLoop(states) {
+    this.states = states;
```

Comment:
This doesn't allow for nested loops. I'm pretty sure nested loops are going to be a bit too complex for what we're doing, but as implemented it actually _allows_ nested loops, they just won't work right.

To make it clearer here, I'd set `this.states` to null when not in a for loop, and check here that it's null and otherwise throw an error (instead of overwriting the state with no error).